File: minting/src/main/scala/co/topl/minting/LeaderElectionMinting.scala
package co.topl.minting

import cats.Monad
import cats.implicits._
import co.topl.algebras.Stats
import co.topl.consensus.algebras.LeaderElectionValidationAlgebra
import co.topl.minting.algebras.LeaderElectionMintingAlgebra.VrfHit
import co.topl.minting.algebras.{LeaderElectionMintingAlgebra, VrfProofAlgebra}
import co.topl.models._
import co.topl.models.utility.Ratio
import co.topl.numerics.implicits._
import co.topl.typeclasses.implicits._
import io.circe.Json
import io.circe.syntax._

object LeaderElectionMinting {

  object Eval {

    def make[F[_]: Monad](
      vk:                   VerificationKeys.VrfEd25519,
      thresholdInterpreter: LeaderElectionValidationAlgebra[F],
      vrfProofAlgebra:      VrfProofAlgebra[F],
      statsInterpreter:     Stats[F],
      statsName:            String
    ): LeaderElectionMintingAlgebra[F] = new LeaderElectionMintingAlgebra[F] {

      private def buildHit(slot: Slot, eta: Eta, testProof: Proofs.Knowledge.VrfEd25519, threshold: Ratio): F[VrfHit] =
        VrfHit(
          EligibilityCertificate(testProof, vk, threshold.typedEvidence.evidence, eta),
          slot,
          threshold
        ).pure[F]

      def getHit(relativeStake: Ratio, slot: Slot, slotDiff: Long, eta: Eta): F[Option[VrfHit]] =
        (
          thresholdInterpreter
            .getThreshold(relativeStake, slotDiff)
            .flatTap(r =>
              statsInterpreter
                .write(statsName, Json.obj("s" -> slot.asJson, "sd" -> slotDiff.asJson, "r" -> r.toBigDecimal.asJson))
            ),
          vrfProofAlgebra.proofForSlot(slot, eta),
          vrfProofAlgebra.rhoForSlot(slot, eta)
        ).tupled
          .flatMap { case (threshold, testProof, rho) =>
            thresholdInterpreter
              .isSlotLeaderForThreshold(threshold)(rho)
              .ifM(
                buildHit(slot, eta, testProof, threshold).map(_.some),
                none[VrfHit].pure[F]
              )
          }
    }
  }
}

File: dart/minting/lib/interpreters/LeaderElectionMinting.dart
import 'package:bifrost_common/models/ratio.dart';
import 'package:bifrost_minting/algebras/LeaderElectionMintingAlgebra.dart';
import 'package:bifrost_minting/algebras/vrf_calculator_algebra.dart';
import 'package:bifrost_minting/models/vrf_hit.dart';
import 'package:fixnum/fixnum.dart';
import 'package:topl_protobuf/consensus/models/eligibility_certificate.pb.dart';

class LeaderElectionMinting extends LeaderElectionMintingAlgebra {
  final List<int> _vrfVK;
  // TODO: LeaderElectionValidation algebra
  final VrfCalculatorAlgebra _vrfProofs;

  LeaderElectionMinting(this._vrfVK, this._vrfProofs);

  @override
  Future<VrfHit?> getHit(
      Ratio relativeStake, Int64 slot, Int64 slotDiff, List<int> eta) async {
    final Ratio threshold = Ratio(); // TODO
    final testProof = await _vrfProofs.proofForSlot(slot, eta);
    final rho = await _vrfProofs.rhoForSlot(slot, eta);
    final isLeader = 2 > 1; // TODO
    if (isLeader) {
      return VrfHit(
        EligibilityCertificate(vrfSig: testProof, vrfVK: _vrfVK, eta: eta),
        slot,
        threshold,
      );
    }
    ;
  }
}
